src/cargo/core/resolver/mod.rs

```diff
@@ -1047,6 +1053,9 @@ impl<'a> Context<'a> {
                        .or_insert(Vec::new());
         if !prev.iter().any(|c| c == summary) {
             self.resolve_graph.push(GraphNode::Add(id.clone()));
+            if let Some(link) = summary.links() {
+                self.links.insert(link.to_owned());
```

Comment:
Could this assert that we're not replacing somethign that was already there?

Reply:
Hm that would indeed be bad! Thinking through how this can happen I think we could trigger this if an older cargo created a lock file with a bunch of new crates (that all have `links` in the manifest) and then a newer cargo tries to use that lock file.

In that case, sounds like we should skip the assert for now!

Reply:
Er well, but if we do actually run into that case the previous project surely couldn't build previously becuase it would have been caught later... 

How about this, could we return an error from here to entirely abort resolution?

Reply:
I think I have that working now.